// @alchemy/aa-ethers/dist/esm/provider-adapter.js
import { BaseSmartContractAccount, SmartAccountProvider, getChain, } from "@alchemy/aa-core";
import { defineReadOnly } from "@ethersproject/properties";
import { JsonRpcProvider } from "@ethersproject/providers";
import { AccountSigner } from "./account-signer.js";
export class EthersProviderAdapter extends JsonRpcProvider {
    constructor(opts) {
        super();
        Object.defineProperty(this, "accountProvider", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "withPaymasterMiddleware", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: (overrides) => {
                this.accountProvider.withPaymasterMiddleware(overrides);
                return this;
            }
        });
        Object.defineProperty(this, "withGasEstimator", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: (override) => {
                this.accountProvider.withGasEstimator(override);
                return this;
            }
        });
        Object.defineProperty(this, "withFeeDataGetter", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: (override) => {
                this.accountProvider.withFeeDataGetter(override);
                return this;
            }
        });
        Object.defineProperty(this, "withCustomMiddleware", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: (override) => {
                this.accountProvider.withCustomMiddleware(override);
                return this;
            }
        });
        if ("accountProvider" in opts) {
            this.accountProvider = opts.accountProvider;
        }
        else {
            const chain = getChain(opts.chainId);
            this.accountProvider = new SmartAccountProvider(opts.rpcProvider, opts.entryPointAddress, chain);
        }
    }
    send(method, params) {
        return this.accountProvider.request({ method, params });
    }
    connectToAccount(fn) {
        defineReadOnly(this, "accountProvider", this.accountProvider.connect(fn));
        return this.getAccountSigner();
    }
    getAccountSigner() {
        return new AccountSigner(this);
    }
    getPublicErc4337Client() {
        return this.accountProvider.rpcClient;
    }
    static fromEthersProvider(provider, entryPointAddress) {
        return new EthersProviderAdapter({
            rpcProvider: provider.connection.url,
            entryPointAddress,
            chainId: provider.network.chainId,
        });
    }
}
